_posts/2017-02-19-simplified-database-migrations-in-go.md

```diff
@@ -0,0 +1,98 @@
+* Rollback: an action in which you take your database form one state to another
+
+The observant reader will notice that the definition for both actions are quite similar, or wait, they are the same. This basically means that instead of having a rollback you can just write a new rollforward if you are in a situation where a database update has gone wrong. 
+
+Another reason is that database roll backs are complicated and there are a lot of things that you need to consider, which you most likely won't, when you're writing a rollback. Let me show an example of something that most people would do when writing a new migration adding a new column:
+
+1. Write migrate up that adds column X
+2. Write migrate down that removes column X
+3. Runs migration
+
+So far so good. If step 3 fails you can just rollback right away, no stress, but this could also be solved by using a transaction when you run the migration. A more complicated scenario, and probably more realistic one as well, is that migration went well but a couple of hours later the database starts to act weird. What do you do? Should you run the rollback? Probably not. At this point in time you have already gotten some data in the new column X which you most likely want to keep, so the solution for rolling back isn't just to remove what you just added. Instead you should write a new migration where you define what you should do to not lose any data.
+
+Not everyone agrees with always rolling forward, and that is ok, but this is the way we at Unacast think of migrations.
+
+## Why couldn't we just use an existing migration framework
+
+In regard of the rollback issue we could most likely have used an existing framework but not used the rollback feature. The main issue to why we wrote our own framework was that none of the framework I could find supported reading migration scripts from assets. In basically all our go application we compile everything to one file, meaning that sql files will have a pre-compile step where we generate go files from the content in the sql scripts. At the same time, it would be nice to use a file backed approach if we would like that in the future. This together with our opinionated view of migrations made us write a new migration framework.
+
+## Implementing a migration framework
+
+Before implementing something, it is always good to think about what you want to achieve. The minimum list of features for this project was:
+
+* Run migrations from assets
+* Run migrations from files
+* Run all migrations, that hasn't been run before, in one transaction
+* It should use the existing `sql.DB`
+* Only roll forward
+
+That doesn't sound so hard, and it isn't as we will see. The last point actually makes things easier since it is a cut in scope compared to most migrations framework. When we now the set of features we need to think about the main flow through the code, and it turns out it is quite simple for migrations:
```

Comment:
👍 ✅ ✅ ✅ ✅ 

Changed the sentence with "it is a cut in scope" to: "he last point actually makes things a little bit easier since we are leaving out one feature, rolling back, compared to most other migrations framework."